### PR TITLE
TDS packet splitting support

### LINKED ISSUE #758: MSSQL support packet size configuration
The server sends an info token with the desired packet size.

Currently we send messages in a single packet, regardless of the desired packet size.

### PR DESCRIPTION
**Pull request by tsegismont**
Fixes #758
    
When the client initializes the connection (LOGIN7), it sends a desired size for TDS packets.
    
The server may or may not accept it. In the response, it will give the packet size it wants and we must enforce it. This value may changed with an ENVCHANGE token later.

### CODE REVIEW

**Comment by tsegismont** [comment]
@vietj can you please review? The PR is divided in two commits that can be reviewed independently:

1. refactoring of TDS codec
2. TDS packet splitting support

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java [hunk 1]
```diff
@@ -119,6 +117,30 @@ public MSSQLConnectOptions addProperty(String key, String value) {
...
+   * Get the desired size for TDS packets.
+   *
+   * @return the desired packet size
+   */
+  public int getPacketSize() {
+    return packetSize;
+  }
+
+  /**
+   * Set the desired size for TDS packets.
```
Comment (vietj) [inline vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java:130]: in bytes

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/ExtendedQueryCommandBaseCodec.java [hunk 2]
```diff
@@ -44,158 +36,76 @@ void encode(TdsMessageEncoder encoder) {
...
+    super.handleResultSetDone();
     rowCount = 0;
   }
 
-  protected abstract void handleMessageDecoded();
+  @Override
+  protected void handleReturnValue(ByteBuf payload) {
+    MSSQLPreparedStatement ps = (MSSQLPreparedStatement) cmd.preparedStatement();
+    if (ps.handle == 0) {
+      payload.skipBytes(2); // skip ordinal position
```
Comment (vietj) [inline vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/ExtendedQueryCommandBaseCodec.java:55]: consider calling this a single time, it looks cute but in practice it costs (it falls in the same area as `write`)

Comment (vietj) [reply vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/ExtendedQueryCommandBaseCodec.java:55]: e.g you could do something like

```java
int bytesToSkip = 2 // sky ordinal position
                           + 2 * payload.readUnsignedByte() // skip param name
                           ...;
```

**Comment by vietj** [comment]
are you able to consistently test this behavior ?

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java (hunk 1 above)
Comment (tsegismont) [reply vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java:130]:
```suggestion
   * Set the desired size (in bytes) for TDS packets.
```

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java [hunk 3]
```diff
@@ -119,6 +117,30 @@ public MSSQLConnectOptions addProperty(String key, String value) {
     return (MSSQLConnectOptions) super.addProperty(key, value);
   }
 
+  /**
+   * Get the desired size for TDS packets.
```
Comment (tsegismont) [inline vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java:121]:
```suggestion
   * Get the desired size (in bytes) for TDS packets.
```

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java [hunk 4]
```diff
@@ -119,6 +117,30 @@ public MSSQLConnectOptions addProperty(String key, String value) {
...
+    return packetSize;
+  }
+
+  /**
+   * Set the desired size for TDS packets.
+   * <p>
+   * The client will use the value as a parameter in the LOGIN7 packet.
+   * The server may or may not accept it.
+   *
+   * @param packetSize the desired packet size
```
Comment (tsegismont) [inline vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java:135]:
```suggestion
   * @param packetSize the desired packet size (in bytes)
```

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/ExtendedQueryCommandBaseCodec.java (hunk 2 above)
Comment (tsegismont) [reply vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/ExtendedQueryCommandBaseCodec.java]: Done

**Comment by tsegismont** [comment]
@vietj PTAL

> are you able to consistently test this behavior ?

In `MSSQLMultipleTdsPacketsTest` there's a test that executes a large batch query. The payload is too large to fit into a single packet of default size. Same for the result.

**Comment by vietj** [comment]
that is what I thought but I was not sure.

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java [hunk 5]
```diff
@@ -130,6 +128,30 @@ public MSSQLConnectOptions addProperty(String key, String value) {
...
+   * Set the desired size (in bytes) for TDS packets.
+   * <p>
+   * The client will use the value as a parameter in the LOGIN7 packet.
+   * The server may or may not accept it.
+   *
+   * @param packetSize the desired packet size (in bytes)
+   * @return a reference to this, so the API can be used fluently
+   * @throws IllegalArgumentException if {@code packetSize} is smaller than {@link #MIN_PACKET_SIZE} or bigger than {@link #MAX_PACKET_SIZE}
+   */
+  public MSSQLConnectOptions setPacketSize(int packetSize) {
```
Comment (vietj) [inline vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java:150]: missing bound check advocated by the `@throws` clause

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/MSSQLConnectionFactory.java [hunk 6]
```diff
@@ -28,13 +30,16 @@
 
 public class MSSQLConnectionFactory extends ConnectionFactoryBase {
 
+  private int packetSize;
```
Comment (vietj) [inline vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/MSSQLConnectionFactory.java:33]: final

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/ExtendedBatchQueryCommandCodec.java [hunk 7]
```diff
@@ -26,22 +26,22 @@
...
       return;
     }
-    super.encode(encoder);
+    super.encode();
   }
 
   @Override
-  protected void handleMessageDecoded() {
+  protected void handleDecodingComplete() {
     if (paramsList.size() == 1 || ++messageDecoded == 2) {
```
Comment (vietj) [inline vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/ExtendedBatchQueryCommandCodec.java:45]: ++ will only be applied when `paramsList.size() != -1` is that the correct thing that is to do here ? i.e is it equivalent to

```java
if (paramsList.size() == 1) {
   ...
} else {
  messageDecoded++;
  if (messageDecoded == 2) {
    ...
  } else {
    ...
  }
}
```

Comment (vietj) [reply vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/ExtendedBatchQueryCommandCodec.java:45]: it is not part of this review but a general remark when seing this

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/QueryCommandBaseCodec.java [hunk 8]
```diff
@@ -18,53 +18,40 @@
...
 
+  @Override
   protected void handleNbcRow(ByteBuf payload) {
     rowResultDecoder.handleNbcRow(rowResultDecoder.desc.columnDatas.length, payload);
   }
 
-  protected void handleResultSetDone(int affectedRows) {
+  @Override
+  protected void handleAffectedRows(long count) {
+    rowCount += count;
```
Comment (vietj) [inline vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/QueryCommandBaseCodec.java:51]: we could check `count >= 0` to catch programming error at this point

**Review by vietj — CHANGES_REQUESTED** [review]
A few comments

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/ExtendedBatchQueryCommandCodec.java (hunk 7 above)
Comment (tsegismont) [reply vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/ExtendedBatchQueryCommandCodec.java:45]: Yes, it is on purpose and equivalent to the snippet above.

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/QueryCommandBaseCodec.java (hunk 8 above)
Comment (tsegismont) [reply vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/QueryCommandBaseCodec.java:51]: There are a lot of other assertions we could add. As an example, when decoding parameters, we could check that scale/length are consistent with the `INT` type. I thought about it but couldn't see the value add.

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java (hunk 5 above)
Comment (tsegismont) [reply vertx-mssql-client/src/main/java/io/vertx/mssqlclient/MSSQLConnectOptions.java:150]: Done

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/MSSQLConnectionFactory.java (hunk 6 above)
Comment (tsegismont) [reply vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/MSSQLConnectionFactory.java:33]: Done

**Comment by tsegismont** [comment]
@vietj PTAL

**Review by vietj — APPROVED** [review]

**Pull request merged**